processor/servicegraphprocessor/processor.go

```diff
@@ -288,7 +299,25 @@ func (p *serviceGraphProcessor) onExpire(e *store.Edge) {
 		zap.String("connection_type", string(e.ConnectionType)),
 		zap.Stringer("trace_id", e.TraceID),
 	)
-	stats.Record(context.Background(), statExpiredEdges.M(1))
+
+	if p.config.VirtualNodeFeatureEnabled {
+		// speculate virtual node before edge get expired.
+		if len(e.ClientService) == 0 {
+			e.ClientService = "user"
```

Comment:
I'm a bit worried about this one. Basically, if the feature is enabled, any trace with a server span and no client span will end up marked as `virtual_node`, when it may not be the case. The client span could just be missing, or arriving late.

We could add some logic to check if the server span is an orphan—ie. it doesn't have a parent/client span. In that case, it's fair to assume the client is from _the outside_. What do you think?

Reply:
You're right, but I think there will be more complex processing logic involved, and we can leave it to the next PR or implement it after more thought.

Reply:
Alright, that's fair. I'll open an issue to track this work. In the meantime, I'd add a TODO or comment to be explicit.